openshift/odo
# create url during push with infomation defined in devfile endpoint

**What type of PR is this?**
/area devfile
/area url

**What does does this PR do / why we need it**:
This PR implements the endpoints entry from devfile v2 container definition.
`odo push` creates url without running `odo url create` with the information specified in devfile endpoint.
 
odo push against OpenShift creates resources based on described mappings with Route
odo push against Kubernetes will create Services based on described mappings. If host information is missing in env.yaml a warning message will be shown.


This PR only implements the url creation. 
For url list and describe, need to rewrite the listing logic,  issue https://github.com/openshift/odo/issues/3615 is created to track.
For UDP service support, since service of type clusterIP cannot support both TCP and UDP. Our current logic needs to be change, will work with a separate issue: https://github.com/openshift/odo/issues/3663

`--secure` flag is removed  from `odo url create` to avoid conflict between `env.yaml` and `devfile.yaml`. We will always use `secure` defined in `devfile.yaml`.  The `secure` field will be completely removed within the list/describe issue (#3615 ). 

Added `path` in route & ingress creation. 

For `exposure` and `protocol` defined in devfile, follows behavior defined in devfile 2.0 schema: https://devfile.github.io/devfile/_attachments/api-reference.html

For same port, if `urlName` in env.yaml is different from `devfile.yaml`, will use the name from env.yaml; since env.yaml is considered as a "local" configuration, user runs `odo create` cli to intentionally set the value, therefore env.yaml has precedence over devfile.yaml 


Also implement https://github.com/openshift/odo/issues/3718#issuecomment-670648628, for URLs for devfile components, the ingress and routes created will be `urlName-componentName`. It won't affect the name saved in env.yaml file, also `url list`, `url describe` and `url, delete` will still use `urlName`, so user won't get confused. 



**Which issue(s) this PR fixes**:

Fixes #3515 

**PR acceptance criteria**:




**How to test changes / Special notes to the reviewer**:

`make test-cmd-devfile-url`


Manual test:

Against Openshift cluster:
run `odo create` with the `devfile-with-multiple-endpoints.yaml` in this pr

directlry run `odo push`
should see 1 service with port 8080, 8090, 9090 created. 
should see 3 routes got created.
run `oc get route <routename> -o yaml`, should see route `8090-tcp` with path `/testpath`
 should see route `9090-tcp` is secured

run `odo url create testurl1 --port 8080 --host com --ingress`
modify `devfile`, remove endpoint entry with port=8080 (or can change the port, or can add `exposure: internal`)and run `odo push`. Should see error that port 8080 is not exposed. 

Modify the devfile endpoint entry, add `exposure: none` to one of the endpoints. 
`odo push` should see service does not contain the corresponding port
